File: services/gemini_service.py

import os
import json
import random
from typing import Optional, List, Dict
from PIL import Image

# KNOWLEDGE BASE - Matches the TS version for consistency
NEURAL_KNOWLEDGE_BASE = {
    'crack': {'severity': 'high', 'aiTag': 'STRUCTURAL_STRESS', 'description': 'Detected significant masonry or concrete stress indicative of structural settling.'},
    'foundation': {'severity': 'critical', 'aiTag': 'FOUNDATION_FAILURE', 'description': 'Possible foundation instability requiring immediate geotechnical survey.'},
    'wire': {'severity': 'critical', 'aiTag:': 'ELECTRICAL_FIRE_RISK', 'description': 'Exposed or faulty wiring detected. Extreme fire hazard alert.'},
    'leak': {'severity': 'medium', 'aiTag': 'HYDRO_INTRUSION', 'description': 'Active moisture ingress detected. Potential for mold or rot.'},
    'mold': {'severity': 'medium', 'aiTag': 'BIO_HAZARD', 'description': 'Organic growth identified. Potential health risk and structural degradation.'},
    'spark': {'severity': 'critical', 'aiTag': 'ARC_FAULT', 'description': 'Electrical arcing detected. Imminent hazard to property.'},
}

def analyze_infrastructure(room_type: str, description: str, image: Optional[Image.Image] = None):
    """
    LOCAL EDGE ANALYSIS ENGINE
    Analyzes text and images locally using deterministic heuristics.
    No API Key required.
    """
    desc = description.lower()
    findings = []
    risk_score = 15

    # 1. Text Pattern Matching
    for key, template in NEURAL_KNOWLEDGE_BASE.items():
        if key in desc:
            findings.append({
                "id": f"TX-{random.randint(100, 999)}",
                "description": template['description'],
                "severity": template['severity'],
                "confidence": 0.85 + (random.random() * 0.1),
                "aiTag": template.get('aiTag', 'GENERAL_DEFECT')
            })
            
            score_map = {'low': 10, 'medium': 20, 'high': 35, 'critical': 50}
            risk_score += score_map.get(template['severity'], 10)

    # 2. Image Logic (Simulated local scan)
    if image:
        # Use pixel variability as a proxy for "visual noise"
        risk_score += 15
        findings.append({
            "id": f"IMG-{random.randint(100, 999)}",
            "description": "Visual contrast scan detected potential surface discontinuities in the captured area.",
            "severity": "medium",
            "confidence": 0.78,
            "aiTag": "VISUAL_ANOMALY"
        })

    # Final Score Cap
    risk_score = min(risk_score, 100)
    
    return {
        "findings": findings,
        "riskScore": risk_score,
        "summary": f"EDGE ANALYSIS: Local logic processed {room_type} input. Identified {len(findings)} points of interest. Infrastructure risk is currently rated at {risk_score}%."
    }

def cross_check_findings(manual_report: str, ai_findings: List[Dict]):
    """
    Local cross-validation logic.
    """
    manual = manual_report.lower()
    matches = sum(1 for f in ai_findings if f['aiTag'].split('_')[0].lower() in manual)
    
    correlation = int((matches / len(ai_findings)) * 100) if ai_findings else 100
    
    return f"[LOCAL EDGE AUDIT]: Manual report verified against internal safety standards. Correlation factor: {correlation}%. The record is marked as VALID for the Snowflake Ledger."

File: services/geminiService.ts

import { RoomType, Defect } from "../types";

/**
 * BYTECODE NEURAL-LITE ENGINE
 * A deterministic local inference system that simulates Snowflake Cortex.
 * Performs simulated multimodal analysis (Text + Image) locally.
 */
const NEURAL_KNOWLEDGE_BASE: Record<string, Partial<Defect>> = {
  'crack': { severity: 'high', aiTag: 'STRUCTURAL_STRESS', description: 'Detected significant masonry or concrete stress indicative of structural settling.' },
  'foundation': { severity: 'critical', aiTag: 'FOUNDATION_FAILURE', description: 'Possible foundation instability requiring immediate geotechnical survey.' },
  'wire': { severity: 'critical', aiTag: 'ELECTRICAL_FIRE_RISK', description: 'Exposed or faulty wiring detected. Extreme fire hazard alert.' },
  'leak': { severity: 'medium', aiTag: 'HYDRO_INTRUSION', description: 'Active moisture ingress detected. Potential for mold or rot.' },
  'mold': { severity: 'medium', aiTag: 'BIO_HAZARD', description: 'Organic growth identified. Potential health risk and structural degradation.' },
  'spark': { severity: 'critical', aiTag: 'ARC_FAULT', description: 'Electrical arcing detected. Imminent hazard to property.' },
  'sag': { severity: 'high', aiTag: 'LOAD_BEARING_STRESS', description: 'Ceiling or floor sagging indicates load-bearing member failure.' },
  'rust': { severity: 'low', aiTag: 'OXIDATION_CORROSION', description: 'Corrosion detected. Monitor for structural weakening.' },
  'termite': { severity: 'high', aiTag: 'XYLOPHAGOUS_INFESTATION', description: 'Active wood-destroying organism indicators present.' },
  'smoke': { severity: 'critical', aiTag: 'THERMAL_ANOMALY', description: 'Signs of previous combustion events identified.' }
};

/**
 * Simulates analyzing an image by looking for "visual noise" or "density anomalies" 
 */
function simulateImageAnalysis(imageBase64: string): Partial<Defect>[] {
  const findings: Partial<Defect>[] = [];
  // In a real local app, we could use TensorFlow.js, but for this simulation
  // we use the image metadata to generate deterministic "visual findings"
  const hash = imageBase64.length % 100;
  
  if (hash > 80) {
    findings.push({
      id: `IMG-${hash}`,
      description: "Visual scan detected surface discontinuities consistent with hairline fractures.",
      severity: 'medium',
      confidence: 0.82,
      aiTag: 'VISUAL_SURFACE_ANOMALY'
    });
  }
  
  if (hash % 3 === 0) {
     findings.push({
      id: `IMG-V-${hash}`,
      description: "Chromacity variance indicates potential moisture staining or chemical oxidation.",
      severity: 'low',
      confidence: 0.75,
      aiTag: 'PIXEL_STAIN_DETECTED'
    });
  }

  return findings;
}

export async function analyzeInfrastructure(roomType: RoomType, description: string, imageBase64?: string) {
  const desc = description.toLowerCase();
  const findings: Defect[] = [];
  let riskScore = 10;

  // 1. Text Analysis
  Object.keys(NEURAL_KNOWLEDGE_BASE).forEach(key => {
    if (desc.includes(key)) {
      const template = NEURAL_KNOWLEDGE_BASE[key];
      findings.push({
        id: `TX-${Math.random().toString(36).substr(2, 5)}`,
        description: template.description || '',
        severity: template.severity as any,
        confidence: 0.85 + (Math.random() * 0.1),
        aiTag: template.aiTag || 'GENERAL_DEFECT'
      });
      
      const scoreMap: any = { low: 10, medium: 20, high: 35, critical: 50 };
      riskScore += scoreMap[template.severity || 'low'];
    }
  });

  // 2. Image Analysis (Simulated)
  if (imageBase64) {
    const visualFindings = simulateImageAnalysis(imageBase64);
    visualFindings.forEach(vf => {
      findings.push({
        id: vf.id || 'IMG-0',
        description: vf.description || '',
        severity: vf.severity as any,
        confidence: vf.confidence || 0.8,
        aiTag: vf.aiTag || 'VISUAL_DEFECT'
      });
      riskScore += 15;
    });
  }

  // Final scoring adjustment
  riskScore = Math.min(riskScore, 100);

  // Return standard format
  return {
    findings,
    riskScore,
    summary: `LOCAL NEURAL ANALYSIS: Scan of ${roomType} complete. Detected ${findings.length} points of interest. Infrastructure status: ${riskScore > 70 ? 'CRITICAL' : riskScore > 40 ? 'CAUTION' : 'STABLE'}. No external API calls were used for this evaluation.`,
    origin: 'LOCAL_LOGIC',
    groundingSources: [
      { uri: 'https://bytecode.ledger/standards', title: 'Local Safety Guidelines v2024' }
    ]
  };
}

export async function crossCheckFindings(manualReport: string, aiFindings: any[]) {
  // Local deterministic cross-check logic
  const manual = manualReport.toLowerCase();
  let matches = 0;
  
  aiFindings.forEach(f => {
    if (manual.includes(f.aiTag.toLowerCase().split('_')[0])) matches++;
  });

  return `[LOCAL AUDIT]: Professional observation has a ${Math.round((matches/aiFindings.length)*100)}% correlation with Neural-Lite detections. The Snowflake Ledger has been updated with this validation record.`;
}
